python/variable_init.cpp

```diff
@@ -0,0 +1,456 @@
+  if (py::isinstance<py::buffer>(obj)) {
+    ensure_is_scalar(obj);
+    return converting_cast<PyType>::cast(obj.attr("item")());
+  } else {
+    return converting_cast<PyType>::cast(obj);
+  }
+}
+
+template <>
+core::time_point extract_scalar<core::time_point>(const py::object &obj,
+                                                  const units::Unit unit) {
+  using TM = ElementTypeMap<core::time_point>;
+  using PyType = typename TM::PyType;
+  TM::check_assignable(obj, unit);
+  if (py::isinstance<py::buffer>(obj)) {
+    ensure_is_scalar(obj);
+    return core::time_point{obj.attr("astype")(py::dtype::of<PyType>())
+                                .attr("item")()
+                                .template cast<PyType>()};
+  } else {
+    return core::time_point{obj.cast<PyType>()};
+  }
+}
+
+template <>
+python::PyObject extract_scalar<python::PyObject>(const py::object &obj,
+                                                  const units::Unit unit) {
+  using TM = ElementTypeMap<python::PyObject>;
+  TM::check_assignable(obj, unit);
+  return obj;
+}
+
+template <class T> struct MakeVariableDefaultInit {
+  static Variable apply(const Dimensions &dims, const units::Unit unit,
+                        const std::optional<bool> with_variance) {
+    auto var = with_variance.value_or(false)
+                   ? makeVariable<T>(Dimensions{dims}, Values{}, Variances{})
+                   : makeVariable<T>(Dimensions{dims});
+    var.setUnit(unit);
+    return var;
+  }
+};
+
+Variable make_variable_default_init(const py::object &dim_labels,
+                                    const py::object &shape,
+                                    const units::Unit unit, DType dtype,
+                                    const std::optional<bool> with_variance) {
+  const auto dims = build_dimensions(dim_labels, shape);
+  if (dtype == core::dtype<void>) {
+    // When there is no way to deduce the dtype, default to double.
+    dtype = core::dtype<double>;
+  }
+  return core::CallDType<
+      double, float, int64_t, int32_t, bool, scipp::core::time_point,
+      std::string, Variable, DataArray, Dataset, Eigen::Vector3d,
+      Eigen::Matrix3d,
+      python::PyObject>::apply<MakeVariableDefaultInit>(dtype, dims, unit,
+                                                        with_variance);
+}
+
+template <class T> struct MakeVariableScalar {
+  static Variable apply(const py::object &value, const py::object &variance,
+                        units::Unit unit,
+                        const std::optional<bool> with_variance) {
+    auto variable = [&]() {
+      auto val = [&]() {
+        if (value.is_none()) {
+          return element_array<T>();
+        } else {
+          const auto [actual_unit, conversion_factor] =
+              common_unit<T>(value, unit);
+          unit = actual_unit;
+          if (conversion_factor != 1) {
+            // TODO Triggered once common_unit implements conversions.
+            std::terminate();
```

Comment:
When does this happend? For datetime? Why terminate?

Reply:
Only for datetime as things are right now. The terminate should be unreachable because `common_unit` throws if `conversion_factor != 1` so it is only a safety feature.
This is only a stop gap solution, I already have an idea how to implement the conversion but I didn't want to add even more feature to this PR.